src/lib.rs

```diff
@@ -227,21 +249,20 @@ impl<T> Arena<T> {
     #[inline]
     pub fn insert(&mut self, object: T) -> usize {
         self.len += 1;
-
-        if self.head == !0 {
-            self.slots.push(Slot::Occupied(object));
-            self.len - 1
-        } else {
-            let index = self.head;
-            match self.slots[index] {
+        let index = self.head;
+        match self.slots.get_mut(index) {
+            None => {}
+            Some(slot) => match mem::replace(slot, Slot::Occupied(object)) {
                 Slot::Vacant(next) => {
                     self.head = next;
-                    self.slots[index] = Slot::Occupied(object);
-                },
+                    return index;
+                }
                 Slot::Occupied(_) => unreachable!(),
             }
-            index
         }
+        self.slots.push(Slot::Occupied(object));
+        self.head = self.len;
+        self.len - 1
```

Comment:
Can these three lines be moved inside the `None => {}` block?
If not, let's change `match self.slots.get_mut(index)` to `if let Some(slot) = self.slots.get_mut(index)`.